File: src/utils/alerts.ts
import type { AlertSeverity, DailyForecast, Friend } from '../types';

export const SEVERITY_LABEL: Record<AlertSeverity, string> = {
  advisory: 'Weather Advisory',
  watch: 'Severe Weather Watch',
  warning: 'Severe Weather Warning',
  emergency: 'Weather Emergency',
};

export const SEVERITY_COLOR: Record<AlertSeverity, string> = {
  advisory: '#38bdf8',
  watch: '#eab308',
  warning: '#f97316',
  emergency: '#ef4444',
};

export function buildAlertMessage(
  day: DailyForecast,
  severity: AlertSeverity,
  customNote: string,
  typeLabel?: string,
): { headline: string; body: string } {
  const dateLabel = new Date(`${day.date}T00:00:00`).toLocaleDateString(undefined, {
    weekday: 'long',
    month: 'short',
    day: 'numeric',
  });

  const headline = `${typeLabel || SEVERITY_LABEL[severity]} for your current location — ${dateLabel}`;

  const lines = [headline];

  if (customNote.trim()) {
    lines.push('', `Note: ${customNote.trim()}`);
  }
  lines.push('', "Sent via Will's Severe Weather Alerts");

  return { headline, body: lines.join('\n') };
}

function encodeParams(params: Record<string, string>): string {
  return Object.entries(params)
    .map(([key, value]) => `${key}=${encodeURIComponent(value)}`)
    .join('&');
}

export function buildSmsLink(friend: Friend, body: string): string {
  return `sms:${friend.phone}?${encodeParams({ body })}`;
}

File: mobile/lib/utils/alerts.dart
// Mirrors src/utils/alerts.ts.
import 'dart:io' show Platform;
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:intl/intl.dart';
import '../models/models.dart';

const Map<AlertSeverity, String> severityLabel = {
  AlertSeverity.advisory: 'Weather Advisory',
  AlertSeverity.watch: 'Severe Weather Watch',
  AlertSeverity.warning: 'Severe Weather Warning',
  AlertSeverity.emergency: 'Weather Emergency',
};

const Map<AlertSeverity, int> severityColor = {
  AlertSeverity.advisory: 0xFF38BDF8,
  AlertSeverity.watch: 0xFFEAB308,
  AlertSeverity.warning: 0xFFF97316,
  AlertSeverity.emergency: 0xFFEF4444,
};

({String headline, String body}) buildAlertMessage(
  String locationName,
  DailyForecast day,
  AlertSeverity severity,
  String customNote,
  String? typeLabel,
) {
  final date = DateTime.parse(day.date);
  final dateLabel = DateFormat('EEEE, MMM d').format(date);

  final headline =
      '${(typeLabel != null && typeLabel.isNotEmpty) ? typeLabel : severityLabel[severity]} for $locationName — $dateLabel';

  final lines = <String>[headline];
  final note = customNote.trim();
  if (note.isNotEmpty) {
    lines.addAll(['', 'Note: $note']);
  }
  lines.addAll(['', "Sent via Will's Severe Weather Alerts"]);

  return (headline: headline, body: lines.join('\n'));
}

/// sms: URIs differ by platform — iOS uses "&body=", Android uses "?body=".
String buildSmsUri(String phone, String body) {
  final encoded = Uri.encodeComponent(body);
  final separator = (!kIsWeb && Platform.isIOS) ? '&' : '?';
  return 'sms:$phone$separator' 'body=$encoded';
}
